src/stream/src/executor/backfill/cdc/cdc_backfill.rs

```diff
@@ -382,33 +466,49 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
+                                            ?current_pk_pos,
+                                            "snapshot read stream ends"
+                                        );
+                                        // If the snapshot read stream ends with less than `limit` rows,
+                                        // it means all historical data has been loaded.
+                                        // We should not mark the chunk anymore,
+                                        // otherwise, we will ignore some rows in the buffer.
+                                        // Here we choose to never mark the chunk.
+                                        // Consume with the renaming stream buffer chunk without mark.
+                                        for chunk in upstream_chunk_buffer.drain(..) {
+                                            yield Message::Chunk(mapping_chunk(
+                                                chunk,
+                                                &self.output_indices,
+                                            ));
+                                        }
 
-                                    state_impl
-                                        .mutate_state(
-                                            current_pk_pos,
-                                            last_binlog_offset.clone(),
-                                            total_snapshot_row_count,
-                                            true,
-                                        )
-                                        .await?;
-                                    break 'backfill_loop;
+                                        state_impl
+                                            .mutate_state(
+                                                current_pk_pos.clone(),
+                                                last_binlog_offset.clone(),
+                                                total_snapshot_row_count,
+                                                true,
+                                            )
+                                            .await?;
+
+                                        // exit backfill
+                                        break 'backfill_loop;
+                                    } else {
+                                        tracing::info!(
+                                            upstream_table_id,
+                                            ?current_pk_pos,
+                                            ?snapshot_read_row_cnt,
+                                            "snapshot stream will contiune"
+                                        );
+
+                                        // break the for loop to reconstruct a new snapshot with pk offset
+                                        // to ensure we load all historical data
+                                        break;
+                                    }
```

Comment:
This break statement will cause different snapshot reads during one epoch. I am not sure whether it would affect correctness in the CDC case. Maybe our table overwrite conflict check could guarantee the correctness. Previously, I thought we would just wait until a next barrier to start a new snapshot read.

Reply:
Sorry for the late reply.
just to check, in this PR's implementation, will the low_watermark be updated again when starting the new snapshot read? If yes, I think it is ok.

Reply:
In the implementation, the binlog_low(last_binlog_offset in code) will be updated to the latest binlog offset among all the events in memory buffer when handle a barrier message.

Reply:
Oh, sorry, I think it is ok to use the binlog_low got before